sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFn.java

```diff
@@ -277,7 +292,9 @@ protected void finalize() {
 
     @Override
     public long estimate() {
-      return memoizedBacklog.get();
+      memoizedBacklog = Preconditions.checkStateNotNull(backlogMap.get(topicPartition));
+      Long backlogValue = memoizedBacklog.get();
```

Comment:
AFAICT this estimate is also called from the main processing thread and thus may still block issuing the backlog request.

Instead we may want to have estimate be non blocking and just return whatever the latest observed offset was, and separately change the background fetching thread to force refreshes of all the backlogs.